src/pubsub/ua_pubsub.c

```diff
@@ -182,6 +185,1098 @@ UA_Server_removeWriterGroup(UA_Server *server, const UA_NodeId writerGroup){
+    UA_StatusCode retval = UA_STATUSCODE_BADUNEXPECTEDERROR;
+    UA_PubSubDataSetReader* pDS = UA_PubSubDataSetReader_findDSRbyId(server, dataSetReaderIdentifier);
+    if (pDS == NULL) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    if (pDS->subscribedDataSetTarget.targetVariablesSize > 0) {
+        UA_TargetVariablesDataType_deleteMembers(&pDS->subscribedDataSetTarget);
+        pDS->subscribedDataSetTarget.targetVariablesSize = 0;
+        pDS->subscribedDataSetTarget.targetVariables = NULL;
+    }
+
+    //Set subscribed dataset to TargetVariableType
+    pDS->subscribedDataSetType = UA_PUBSUB_SDS_TARGET;
+    retval = UA_TargetVariablesDataType_copy(targetVariables, &pDS->subscribedDataSetTarget);
+    return retval;
+}
+
+/**
+ * Adds Subscribed Variables from the DataSetMetaData for the given DataSet into the given parent node
+ * and creates the corresponding data in the targetVariables of the DataSetReader
+ *
+ * @param server
+ * @param parentNode
+ * @param dataSetReaderIdentifier
+ * @return UA_STATUSCODE_GOOD on success
+ */
+UA_StatusCode UA_PubSubDataSetReader_addTargetVariables(UA_Server *server, UA_NodeId *parentNode, UA_NodeId dataSetReaderIdentifier, UA_SubscribedDataSetEnumType sdsType) {
+    if ((server == NULL) || (parentNode == NULL)) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_PubSubDataSetReader* pDataSetReader = UA_PubSubDataSetReader_findDSRbyId(server, dataSetReaderIdentifier);
+    if (pDataSetReader == NULL) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_TargetVariablesDataType targetVars;
+    targetVars.targetVariablesSize = pDataSetReader->config.dataSetMetaData.fieldsSize;
+    targetVars.targetVariables = (UA_FieldTargetDataType *)UA_calloc(targetVars.targetVariablesSize, sizeof(UA_FieldTargetDataType));
+    UA_StatusCode retVal = UA_STATUSCODE_GOOD;
+    for (size_t iteratorField = 0; iteratorField < pDataSetReader->config.dataSetMetaData.fieldsSize; iteratorField++) {
+        UA_VariableAttributes vAttr = UA_VariableAttributes_default;
+        vAttr.valueRank = pDataSetReader->config.dataSetMetaData.fields[iteratorField].valueRank;
+        if (pDataSetReader->config.dataSetMetaData.fields[iteratorField].arrayDimensionsSize > 0) {
+            retVal = UA_Array_copy(pDataSetReader->config.dataSetMetaData.fields[iteratorField].arrayDimensions, pDataSetReader->config.dataSetMetaData.fields[iteratorField].arrayDimensionsSize, (void**)&vAttr.arrayDimensions, &UA_TYPES[UA_TYPES_UINT32]);
+            if (retVal == UA_STATUSCODE_GOOD) {
+                vAttr.arrayDimensionsSize = pDataSetReader->config.dataSetMetaData.fields[iteratorField].arrayDimensionsSize;
+            }
+        }
+
+        switch (pDataSetReader->config.dataSetMetaData.fields[iteratorField].builtInType) {
```

Comment:
There is a lot of code duplication going on here.

Don't.

Instead, find the correct dataype in the UA_TYPES array.
And then use the same code path for type.